Implement a SQL-backed storage layer for Reporting

cwbriones opened the pull request:
This currently does not map the data model to the internal model, so the interfaces expose the data models right now (but nothing uses them).

Code is mostly boilerplate and supports CRUD operations on a `ReportingJob` instance.

There is currently no notion of an execution history as we discussed, but I figured I can implement that some time after this to keep down the PR size.

vjkoskela commented on app/com/arpnetworking/metrics/portal/reports/ReportingJobRepository.java:
```diff
@@ -0,0 +1,63 @@
+
+package com.arpnetworking.metrics.portal.reports;
+
+// TODO(cwbriones): Decouple from ebean model and expose via the internal model.
```
Would it be possible to include the interfaces for the internal model in this PR and adjust the code to use those?

vjkoskela commented on app/com/arpnetworking/metrics/portal/reports/ReportingJobRepository.java:
```diff
@@ -0,0 +1,63 @@
+import java.util.UUID;
+
+/**
+ * A repository for storing and retrieving instances of <code>ReportingJob</code>.
```
Ideally, we should use `{@link ...}` or `{@code ...}`. The use of `<code>...</code>` is generally frowned upon. Unfortunately, this project has not been updated.

vjkoskela commented on app/models/ebean/ChromeScreenshotReportSource.java:
```diff
@@ -0,0 +1,102 @@
+ */
+// CHECKSTYLE.OFF: MemberNameCheck
+@Entity
+@DiscriminatorValue("chromeScreenshot")
```
This is the "enumerated" type descriptor? Should we use capital snake case to link it to such?

vjkoskela commented on app/models/ebean/ChromeScreenshotReportSource.java:
```diff
@@ -0,0 +1,102 @@
+    @Column(name = "triggering_event_name")
+    private String triggeringEventName;
+
+    @Column(name = "pdf_width_inches")
```
Why does the report source contain format specific flags?

cwbriones replied:
I'll pull this out into formats for the report jobs as per discussion.

vjkoskela commented on app/models/ebean/ReportRecipient.java:
```diff
@@ -0,0 +1,128 @@
+
+    /**
+     * Create a new ReportRecipient with the given emailAddress.
+     * @param emailAddress The address of the recipient
```
nit: blank line between description and tags.

vjkoskela commented on app/models/ebean/ReportRecipient.java:
```diff
@@ -0,0 +1,128 @@
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    @Column(name = "id")
+    private Long id;
```
Existing conventions would have the fields at the top, followed by methods, followed by subtypes.

vjkoskela commented on app/models/ebean/ReportRecipient.java:
```diff
@@ -0,0 +1,128 @@
+
+    @Override
+    public String toString() {
+        return "ReportRecipient{"
```
Use `MoreObjects` to string helper.

vjkoskela commented on app/models/ebean/ReportRecipient.java:
```diff
@@ -0,0 +1,128 @@
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
```
IMO you should use `.equals` for the class equality.

vjkoskela replied:
Alternatively, marking this class `final` and using `instanceof` works nicely here since you don't need a null check.

cwbriones replied:
This was autogenerated by IntelliJ, maybe I can configure what it actually spits out? I'll still update this though.

vjkoskela commented on app/models/ebean/ReportRecipientGroup.java:
```diff
@@ -0,0 +1,121 @@
+ */
+@Entity
+@Table(name = "report_recipient_groups", schema = "portal")
+// CHECKSTYLE.OFF: MemberNameCheck
```
The class above has `// CHECKSTYLE.ON: FinalClassCheck`,  but this does not.

cwbriones replied:
Noted, although I'm not sure why Checkstyle isn't complaining about this instance.

vjkoskela commented on app/models/ebean/ReportRecipient.java:
```diff
@@ -0,0 +1,128 @@
+     * @return The address of the recipient.
+     */
+    public String get() {
+        return recipient;
```
Are there any cases where this is insufficient? Consider common targets such as Slack, S3, Dropbox, etc.

vjkoskela replied:
Should this be discriminated and "polymorphic" as well?

cwbriones replied:
I think this should be sufficient, each of those is described by a URI and customization beyond that wouldn't be a requirement of the sink. (Slack - channel to post to, S3 - bucket, Dropbox - filepath)

I suppose there might be other parameters we might want to associate with it (e.g. dropbox uploads with tags, slack posts with a custom message along with the content itself)

vjkoskela commented on app/models/ebean/ReportSource.java:
```diff
@@ -0,0 +1,96 @@
+        return timeoutInSeconds;
+    }
+
+    public void setTimeoutInSeconds(final long value) {
```
So the assertion is that all report types have a timeout?

cwbriones replied:
Yes. This can arguably moved to the actual job itself, but my thought was you should be able to specify a limit on how long the execution can run. Now that I think about it though, this is another hook in which external users could then influence the scheduler so maybe this is better left off as an internal detail.

vjkoskela commented on app/models/ebean/ReportingSchedule.java:
```diff
@@ -0,0 +1,96 @@
+    private Integer id;
+
+    @Column(name = "send_at")
+    private Timestamp sendAt;
```
Do we want to give the user this much control over scheduling?

vjkoskela commented on app/models/ebean/ReportingSchedule.java:
```diff
@@ -0,0 +1,96 @@
+        /**
+         * A schedule which recurs weekly.
+         */
+        WEEKLY
```
So this union structure will capture all data for all types of schedules? e.g. third Monday of month, weekly from date x until date y, daily from date x for y occurrences, etc.

vjkoskela replied:
Will this table also be updated with data necessary to evaluate the schedule? Or are we going to obtain that elsewhere? (e.g. most recently sent date)

cwbriones replied:
> So this union structure will capture all data for all types of schedules? e.g. third Monday of month, weekly from date x until date y, daily from date x for y occurrences, etc.

I updated this to be polymorphic in the type, from a quick run-through I think that should be sufficient to express any of the complex logic that we might want. (e.g. 3rd monday of the month, weekly on M W F)

> Will this table also be updated with data necessary to evaluate the schedule?

Job executions will be in a separate table, that data could be pulled out from there.

vjkoskela commented on conf/db/migration/metrics_portal_ddl/common/V11__create_reporting_tables.sql:
```diff
@@ -0,0 +1,73 @@
+    timeout_seconds INTEGER NOT NULL,
+    type VARCHAR(255) NOT NULL,
+
+    -- Chrome
```
Is this the actual value of `type` for when these fields apply?

cwbriones replied:
No, that was more for my convenience of tracking which field belonged to which instance. I can rename it though to line up with the discriminator value.

vjkoskela commented on conf/db/migration/metrics_portal_ddl/common/V11__create_reporting_tables.sql:
```diff
@@ -0,0 +1,73 @@
+    ignore_certificate_errors BIT DEFAULT 0,
+    triggering_event_name VARCHAR(255),
+    pdf_width_inches FLOAT,
+    pdf_height_inches FLOAT,
```
Thoughts or patterns for avoiding column collisions between types?

cwbriones replied:
I don't like this either; I wanted to store the properties of the subclasses in their own tables to avoid this problem but Ebean only allows `InheritanceType.SINGLE_TABLE`. I'll give this some more thought, I also don't like that I can't enforce NONNULL with this strategy.

vjkoskela commented on conf/db/migration/metrics_portal_ddl/common/V11__create_reporting_tables.sql:
```diff
@@ -0,0 +1,73 @@
+    id BIGSERIAL PRIMARY KEY,
+    uuid UUID NOT NULL,
+    version INTEGER NOT NULL DEFAULT 1,
+    name VARCHAR(255) DEFAULT '',
```
So is the idea to have full on group management? Or does each report create its own "group" (just wondering what the role of name is)

cwbriones replied:
The name might not be precise enough, I wanted to encapsulate who the report is sent to, rather than storing it as a "Sink" since that concept only makes sense to me at runtime. 

Really all this is trying to be is a set of recipients for a single report, with all of the individual addresses in an umbrella underneath it.

vjkoskela commented on conf/db/migration/metrics_portal_ddl/common/V11__create_reporting_tables.sql:
```diff
@@ -0,0 +1,73 @@
+CREATE UNIQUE INDEX report_recipient_groups_uuid_idx ON portal.report_recipient_groups (uuid);
+CREATE UNIQUE INDEX report_recipient_groups_name_idx ON portal.report_recipient_groups (name);
+
+CREATE TABLE portal.report_recipients (
```
should this be `report_group_members`?

vjkoskela commented on test/com/arpnetworking/metrics/portal/TestBeanFactory.java:
```diff
@@ -44,6 +51,49 @@
  */
 public final class TestBeanFactory {
 
+    private static final String TEST_EMAIL = "noreply+email-recipient@test.com";
```
All these should be randomly generated or at least randomly selected from sets where generation is not possible.

cwbriones commented on app/com/arpnetworking/metrics/portal/reports/impl/DatabaseReportRepository.java:
```diff
@@ -0,0 +1,423 @@
+                .where()
+                .eq("report.uuid", reportId)
+                .eq("report.organization.uuid", organization.getId())
+                .eq("state", ReportExecution.State.SUCCESS)
```
Ebean uses the `@SoftDelete` annotation to mark a field as soft-deleted and automatically filters those out by default unless you include `includeSoftDeleted()` in your query.

cwbriones commented on app/com/arpnetworking/metrics/portal/reports/impl/DatabaseReportRepository.java:
```diff
@@ -0,0 +1,423 @@
+    @Override
+    public void jobSucceeded(final UUID reportId, final Organization organization, final Instant scheduled, final Report.Result result) {
+        assertIsOpen();
+        if (!(result instanceof DefaultReportResult)) {
```
You're right, I've addressed this by configuring the Result with a `JsonTypeInfo` annotation.

cwbriones commented:
I've gone ahead and addressed most of your feedback, one question about groups:

We've discussed and settled upon the idea that a group can have multiple types of recipients, that is, the internal model should be pretty much the same as the data model here. But the internal model as it stands separates groups into homogenous collections, (e.g. `EmailRecipientGroup`) Should we just leave this as-is for now and update later on? It would be less than ideal but there's quite a bit to untangle there if we do want to change that.

As far as recipient encoding, I still think URIs should suffice. Slack is a channel id + a workspace, and this is already encoded in their URLs. Even if not, I don't think it would be much effort to move this to a JSON encoding later on.

cwbriones commented:
@speezepearson - I think we discussed having `NeverSchedule` be a testing-only type? I don't think we'd want users creating actual production reports that never run.

vjkoskela commented on app/models/ebean/Recipient.java:
```diff
@@ -0,0 +1,192 @@
+    @Column(name = "updated_at")
+    private Timestamp updatedAt;
+
+    @OneToMany(
```
This should not be necessary and couples generic recipients to all consumptions of them (e.g. reports, alerts, etc.)

cwbriones commented:
rebased with master.
